log/loki.go

```diff
@@ -0,0 +1,422 @@
+
+// TODO move this to it's own package
+// benchmark it
+
+type lokiHook struct {
+	addr             string
+	additionalParams [][2]string
+	ch               chan *logrus.Entry
+	limit            int
+	levels           []logrus.Level
+	pushPeriod       time.Duration
+	client           *http.Client
+}
+
+// LokiFromConfigLine returns a new logrus.Hook that pushes logrus.Entrys to loki and is configured
+// through the provided line
+func LokiFromConfigLine(line string) (logrus.Hook, error) {
+	h := &lokiHook{
+		addr:       "http://127.0.0.1:3100/loki/api/v1/push",
+		limit:      100,
+		levels:     logrus.AllLevels,
+		pushPeriod: time.Second * 1, // TODO configurable,
+		client: &http.Client{
+			Timeout: time.Second * 5, // TODO configurable, equal to pushPeriod?
+		},
+	}
+	if line == "loki" {
+		return h, nil
+	}
+
+	parts := strings.SplitN(line, "=", 2)
+	if parts[0] != "loki" {
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
+	if len(args) == 1 {
+		return h, nil
+	}
+
+	for _, arg := range args[1:] {
+		paramParts := strings.SplitN(arg, "=", 2)
+
+		if len(paramParts) != 2 {
+			return nil, fmt.Errorf("loki arguments should be in the form `address,key1=value1,key2=value2`, got %s", arg)
+		}
+
+		key := paramParts[0]
+		value := paramParts[1]
+		switch key {
+		case "additionalParams":
+			values := strings.Split(value, ";") // ; because , is already used
+
+			h.additionalParams = make([][2]string, len(values))
+			for i, value := range values {
+				additionalParamParts := strings.SplitN(value, "=", 2)
+				if len(additionalParamParts) != 2 {
+					return nil, fmt.Errorf("additionalparam should be in the form key1=value1;key2=value2, got %s", value)
+				}
+				h.additionalParams[i] = [2]string{additionalParamParts[0], additionalParamParts[1]}
+			}
+		case "limit":
+			var err error
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+		case "level":
+			// TODO figure out if `tracing`,`fatal` and `panic` should be included
+			h.levels = []logrus.Level{}
+			switch value {
+			case "debug":
+				h.levels = append(h.levels, logrus.DebugLevel)
+				fallthrough
+			case "info":
+				h.levels = append(h.levels, logrus.InfoLevel)
+				fallthrough
+			case "warning":
+				h.levels = append(h.levels, logrus.WarnLevel)
+				fallthrough
+			case "error":
+				h.levels = append(h.levels, logrus.ErrorLevel)
+			default:
+				return nil, fmt.Errorf("unknown log level %s", value)
+			}
+		default:
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.start()
+	return h, nil
+}
+
+func (h *lokiHook) start() {
+	h.ch = make(chan *logrus.Entry, 1000)
```

Comment:
if there's a better (i.e. non manually-parsed) configuration, this should probably be configurable as a `bufferSize` parameter or something like that

Reply:
We can have it even now :D